src/backend/dx11/src/lib.rs

```diff
@@ -93,408 +127,1163 @@ impl Instance {
+                    min_buffer_copy_offset_alignment: 0,    // TODO
+                    min_buffer_copy_pitch_alignment: 0,     // TODO
+                    min_texel_buffer_offset_alignment: 0,   // TODO
+                    min_uniform_buffer_offset_alignment: 0, // TODO
+                    min_storage_buffer_offset_alignment: 0, // TODO
+                    framebuffer_color_samples_count: 0,     // TODO
+                    framebuffer_depth_samples_count: 0,     // TODO
+                    framebuffer_stencil_samples_count: 0,   // TODO
+                    non_coherent_atom_size: 0,              // TODO
+                },
+                memory_properties: hal::MemoryProperties {
+                    memory_types: vec![], // TODO
+                    memory_heaps: vec![]  // TODO
                 }
-            );
+            };
 
-            cur_index += 1;
+            adapters.push(hal::Adapter {
+                info,
+                physical_device,
+                queue_families: vec![QueueFamily]
+            });
         }
 
         adapters
     }
 }
 
-#[derive(Copy, Clone, PartialEq, Eq, Hash, Debug)]
-pub struct Buffer(native::Buffer);
-impl Buffer {
-    pub fn as_resource(&self) -> *mut winapi::ID3D11Resource {
-        type Res = *mut winapi::ID3D11Resource;
-        match self.0 {
-            native::Buffer(t) => t as Res,
+#[derive(Derivative)]
+#[derivative(Debug)]
+pub struct PhysicalDevice {
+    #[derivative(Debug="ignore")]
+    adapter: ComPtr<dxgi::IDXGIAdapter1>,
+    features: hal::Features,
+    limits: hal::Limits,
+    memory_properties: hal::MemoryProperties,
+}
+
+unsafe impl Send for PhysicalDevice { }
+unsafe impl Sync for PhysicalDevice { }
+
+// TODO: does the adapter we get earlier matter for feature level?
+fn get_feature_level(adapter: *mut dxgi::IDXGIAdapter) -> d3dcommon::D3D_FEATURE_LEVEL {
+    let requested_feature_levels = [
+        d3dcommon::D3D_FEATURE_LEVEL_11_1,
+        d3dcommon::D3D_FEATURE_LEVEL_11_0,
+        d3dcommon::D3D_FEATURE_LEVEL_10_1,
+        d3dcommon::D3D_FEATURE_LEVEL_10_0,
+        d3dcommon::D3D_FEATURE_LEVEL_9_3,
+        d3dcommon::D3D_FEATURE_LEVEL_9_2,
+        d3dcommon::D3D_FEATURE_LEVEL_9_1,
+    ];
+
+    let mut feature_level = d3dcommon::D3D_FEATURE_LEVEL_9_1;
+    let hr = unsafe {
+        d3d11::D3D11CreateDevice(
+            adapter,
+            d3dcommon::D3D_DRIVER_TYPE_NULL,
+            ptr::null_mut(),
+            0,
+            requested_feature_levels[..].as_ptr(),
+            requested_feature_levels.len() as _,
+            d3d11::D3D11_SDK_VERSION,
+            ptr::null_mut(),
+            &mut feature_level as *mut _,
+            ptr::null_mut()
+        )
+    };
+
+    if !winerror::SUCCEEDED(hr) {
+        // if there is no 11.1 runtime installed, requesting
+        // `D3D_FEATURE_LEVEL_11_1` will return E_INVALIDARG so we just retry
+        // without that
+        if hr == winerror::E_INVALIDARG {
```

Comment:
what about different errors?

Reply:
According to [msdn](https://msdn.microsoft.com/en-us/library/windows/desktop/ff476082(v=vs.85).aspx) that's the only relevant return code here (we dont request debug)

Reply:
let's assert on that knowledge then ;)